Remotion/Data/DomainObjects.UnitTests/Persistence/ReadOnlyStorageProviderDecoratorTest.cs

```diff
@@ -0,0 +1,131 @@
+
+      Assert.That(() => readOnlyStorageProviderDecorator.ExecuteCustomQuery(queryStub.Object), Throws.Nothing);
+
+      innerReadOnlyStorageProviderStub.Verify(_ => _.ExecuteCustomQuery(queryStub.Object), Times.Once);
+    }
+
+    [Test]
+    [TestCase(QueryType.CollectionReadOnly)]
+    [TestCase(QueryType.CustomReadOnly)]
+    [TestCase(QueryType.ScalarReadOnly)]
+    public void ExecuteScalarQuery_WithReadOnlyQueryType_CallsInnerStorageProvider (QueryType queryType)
+    {
+      var queryStub = new Mock<IQuery>();
+      queryStub.Setup(_ => _.QueryType).Returns(queryType);
+      var innerReadOnlyStorageProviderStub = new Mock<IReadOnlyStorageProvider>();
+      var readOnlyStorageProviderDecorator = new ReadOnlyStorageProviderDecorator(innerReadOnlyStorageProviderStub.Object);
+
+      Assert.That(() => readOnlyStorageProviderDecorator.ExecuteScalarQuery(queryStub.Object), Throws.Nothing);
+
+      innerReadOnlyStorageProviderStub.Verify(_ => _.ExecuteScalarQuery(queryStub.Object), Times.Once);
+    }
+
+    [Test]
+    [TestCase(QueryType.CollectionReadWrite)]
+    [TestCase(QueryType.CustomReadWrite)]
+    [TestCase(QueryType.ScalarReadWrite)]
+    public void ExecuteCollectionQuery_WithReadWriteQueryType_ThrowsArgumentException (QueryType queryType)
+    {
+      var queryStub = new Mock<IQuery>();
+      queryStub.Setup(_ => _.QueryType).Returns(queryType);
+      var innerReadOnlyStorageProviderStub = new Mock<IReadOnlyStorageProvider>();
+      var readOnlyStorageProviderDecorator = new ReadOnlyStorageProviderDecorator(innerReadOnlyStorageProviderStub.Object);
+
+      Assert.That(
+          () => readOnlyStorageProviderDecorator.ExecuteCollectionQuery(queryStub.Object),
+          Throws.InstanceOf<ArgumentException>().With.ArgumentExceptionMessageEqualTo(
+              $"Read-write query '{queryStub.Object}' cannot be used with IReadOnlyStorageProvider.",
+              "query"));
+    }
+
+    [Test]
+    [TestCase(QueryType.CollectionReadWrite)]
+    [TestCase(QueryType.CustomReadWrite)]
+    [TestCase(QueryType.ScalarReadWrite)]
+    public void ExecuteScalarQuery_WithReadWriteQueryType_ThrowsArgumentException (QueryType queryType)
+    {
+      var queryStub = new Mock<IQuery>();
+      queryStub.Setup(_ => _.QueryType).Returns(queryType);
+      var innerReadOnlyStorageProviderStub = new Mock<IReadOnlyStorageProvider>();
+      var readOnlyStorageProviderDecorator = new ReadOnlyStorageProviderDecorator(innerReadOnlyStorageProviderStub.Object);
+
+      Assert.That(
+          () => readOnlyStorageProviderDecorator.ExecuteCollectionQuery(queryStub.Object),
+          Throws.InstanceOf<ArgumentException>().With.ArgumentExceptionMessageEqualTo(
+              $"Read-write query '{queryStub.Object}' cannot be used with IReadOnlyStorageProvider.",
+              "query"));
+    }
+
+    [Test]
+    [TestCase(QueryType.CollectionReadWrite)]
+    [TestCase(QueryType.CustomReadWrite)]
+    [TestCase(QueryType.ScalarReadWrite)]
```

Comment:
you could use a TestCaseSource, supplying an array defined once in a static field. That way, no need to repeat the list of unsupported (or supported) QueryType values.